src/interpreter.ts

```diff
@@ -25,3 +499,840 @@ export function parseAndEvalExpression(sourceCode: string): EvalResult {
+            case "trait":
+                this.interpretTrait(ast);
+                break;
+        }
+    }
+
+    public interpretConstantDef(ast: AstConstantDef) {
+        throwNonFatalErrorConstEval(
+            "Constant definitions are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretFunctionDef(ast: AstFunctionDef) {
+        throwNonFatalErrorConstEval(
+            "Function definitions are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretStructDecl(ast: AstStructDecl) {
+        throwNonFatalErrorConstEval(
+            "Struct declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretMessageDecl(ast: AstMessageDecl) {
+        throwNonFatalErrorConstEval(
+            "Message declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretPrimitiveTypeDecl(ast: AstPrimitiveTypeDecl) {
+        throwNonFatalErrorConstEval(
+            "Primitive type declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretFunctionDecl(ast: AstNativeFunctionDecl) {
+        throwNonFatalErrorConstEval(
+            "Native function declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretContract(ast: AstContract) {
+        throwNonFatalErrorConstEval(
+            "Contract declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretTrait(ast: AstTrait) {
+        throwNonFatalErrorConstEval(
+            "Trait declarations are currently not supported.",
+            ast.loc,
+        );
+    }
+
+    public interpretExpression(ast: AstExpression): Value {
+        switch (ast.kind) {
+            case "id":
+                return this.interpretName(ast);
+            case "method_call":
+                return this.interpretMethodCall(ast);
+            case "init_of":
+                return this.interpretInitOf(ast);
+            case "null":
+                return this.interpretNull(ast);
+            case "boolean":
+                return this.interpretBoolean(ast);
+            case "number":
+                return this.interpretNumber(ast);
+            case "string":
+                return this.interpretString(ast);
+            case "op_unary":
+                return this.interpretUnaryOp(ast);
+            case "op_binary":
+                return this.interpretBinaryOp(ast);
+            case "conditional":
+                return this.interpretConditional(ast);
+            case "struct_instance":
+                return this.interpretStructInstance(ast);
+            case "field_access":
+                return this.interpretFieldAccess(ast);
+            case "static_call":
+                return this.interpretStaticCall(ast);
+        }
+    }
+
+    public interpretName(ast: AstId): Value {
```

Comment:
since Tact does not allow identifier shadowing, `interpretName` should check if there are scoping rules violations. Yes, it could rely on the typechecker to do that, but it's not mature enough at this point, see, for instance, #676

Reply:
I only added checks during let declarations and function calls, because those are the only places where the runtime environment grows at this moment.